ts/Stock/Indicators/Supertrend/SupertrendIndicator.ts

```diff
@@ -0,0 +1,669 @@
+            xData: Array<number> = [],
+            yData: Array<(number|undefined)> = [],
+            close = 3,
+            low = 2,
+            high = 1,
+            periodsOffset = (period === 0) ? 0 : period - 1,
+            basicUp: number,
+            basicDown: number,
+            finalUp: Array<number> = [],
+            finalDown: Array<number> = [],
+            supertrend: (number|undefined),
+            prevFinalUp: number,
+            prevFinalDown: number,
+            prevST: (number|undefined), // previous Supertrend
+            prevY: Array<number>,
+            y: Array<number>,
+            i: number;
+
+        if (
+            (xVal.length <= period) || !isArray(yVal[0]) ||
+            yVal[0].length !== 4 || period < 0
+        ) {
+            return;
+        }
+
+        ATRData = (ATRIndicator.prototype.getValues.call(this, series, {
+            period: period
+        }) as any).yData;
+
+        for (i = 0; i < ATRData.length; i++) {
+            y = yVal[periodsOffset + i];
+            prevY = yVal[periodsOffset + i - 1] || [];
+            prevFinalUp = finalUp[i - 1];
+            prevFinalDown = finalDown[i - 1];
+            prevST = yData[i - 1];
+
+            if (i === 0) {
+                prevFinalUp = prevFinalDown = prevST = 0;
+            }
+
+            basicUp = correctFloat(
+                (y[high] + y[low]) / 2 + multiplier * ATRData[i]
+            );
+            basicDown = correctFloat(
+                (y[high] + y[low]) / 2 - multiplier * ATRData[i]
+            );
+
+            if (
+                (basicUp < prevFinalUp) ||
+                (prevY[close] > prevFinalUp)
+            ) {
+                finalUp[i] = basicUp;
+            } else {
+                finalUp[i] = prevFinalUp;
+            }
+
+            if (
+                (basicDown > prevFinalDown) ||
+                (prevY[close] < prevFinalDown)
+            ) {
+                finalDown[i] = basicDown;
+            } else {
+                finalDown[i] = prevFinalDown;
+            }
+
+            if (prevST === prevFinalUp && y[close] < finalUp[i] ||
+                prevST === prevFinalDown && y[close] < finalDown[i]
+            ) {
+                supertrend = finalUp[i];
+            } else if (
+                prevST === prevFinalUp && y[close] > finalUp[i] ||
+                prevST === prevFinalDown && y[close] > finalDown[i]
+            ) {
+                supertrend = finalDown[i];
+            }
+
+            ST.push([xVal[periodsOffset + i], supertrend]);
+            xData.push(xVal[periodsOffset + i]);
+            yData.push(supertrend);
+        }
+
+        return {
+            values: ST,
+            xData: xData,
+            yData: yData
+        } as IndicatorValuesObject<TLinkedSeries>;
+    }
+}
+
+/* *
+ *
+ *  Prototype Properties
+ *
+ * */
+interface SupertrendIndicator {
+    linkedParent: SupertrendLinkedParentObject;
```

Comment:
`linkedParent` goes into the class pattern (it is not used in the extend call; exception to this rule is `pointClass`)